[11.x] allow guessing of nested component

browner12 opened the pull request:
sometimes components will be part of a larger component grouping, and you may wish to group those components in a folder.

as a simple example we'll use the classic "Card" component, which may be organized as such:

- `App\View\Components\Card\Card`
- `App\View\Components\Card\Header`
- `App\View\Components\Card\Body`

to utilize these in your Blade file you would:

```blade
<x-card.card>
    <x-card.header>Title</x-card.header>
    <x-card.body>lorem ipsum</x-card.body>
</x-card.card>
```

while this is fine, it doesn't read as nice as it could. with this commit, we can now omit the trailing duplicated word, as long as the class name matches the folder name.

```blade
<x-card>
    <x-card.header>Title</x-card.header>
    <x-card.body>lorem ipsum</x-card.body>
</x-card>
```

we do something similar with [Anonymous components](https://laravel.com/docs/11.x/blade#anonymous-index-components)

*If* a component at the root level exists with the same name as the folder, it **will** take precedence.

Tried to write a test for this, but was honestly a little lost.

<!--
Please only send a pull request to branches that are currently supported: https://laravel.com/docs/releases#support-policy 

If you are unsure which branch your pull request should be sent to, please read: https://laravel.com/docs/contributions#which-branch

Pull requests without a descriptive title, thorough description, or tests will be closed.

In addition, please describe the benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.
-->

browner12 committed:
allow guessing of nested component

sometimes components will be part of a larger component grouping, and you may wish to group those components in a folder.

as a simple example we'll use the classic "Card" component, which may be organized as such:

- `App\View\Components\Card\Card`
- `App\View\Components\Card\Header`
- `App\View\Components\Card\Body`

to utilize these in your Blade file you would:

```blade
<x-card.card>
    <x-card.header>Title</x-card.header>
    <x-card.body>lorem ipsum</x-card.body>
</x-card.card>
```

while this is fine, it doesn't read as nice as it could. with this commit, we can now omit the trailing duplicated word, as long as the class name matches the folder name.

```blade
<x-card>
    <x-card.header>Title</x-card.header>
    <x-card.body>lorem ipsum</x-card.body>
</x-card>
```

we do something similar with [Anonymous components](https://laravel.com/docs/11.x/blade#anonymous-index-components)

browner12 committed:
minor formatting

bert-w commented:
This feels like magic; how about creating `App\View\Component\Card.php` so you dont have this issue?

browner12 commented:
@bert-w the point of this is to avoid that, and be able to group you components appropriately. it's the same as what we do with anonymous components.

bert-w commented:
hm this does introduce a conflict however if you have both `App\View\Component\Card.php` and `App\View\Component\Card\Card.php` defined.

taylorotwell commented:
Think we would probably want a test here.

browner12 committed:
add test for nested default component parsing

browner12 commented:
@bert-w correct. I addressed that in my original comment.

someone please check my work on this test. I was not able to figure out how to get it to work with the aliases

browner12 committed:
minor formatting

devajmeireles commented:
@browner12, I have a sincere question! Won't we have some kind of "conflict" with this PR? I mean, your example:

```blade
<x-card>
    <x-card.header>Title</x-card.header>
    <x-card.body>lorem ipsum</x-card.body>
</x-card>
```

... won't it result in the component being understood [as anonymous](https://laravel.com/docs/11.x/blade#anonymous-index-components), possibly invoking `card/index.blade.php` if exists?

browner12 commented:
From how I'm reading the class, the anonymous check happens **after** my check, and all the other class based checks.

Let me know if you're reading it differently.

However, you *might* be able to consider this a BC break (?) under the following scenario:

- `/App/View/Components/Card/Card.php`
- `/resources/views/components/card/index.blade.php`
- `/resources/views/components/card/header.blade.php`
- `/resources/views/components/card/body.blade.php`
- `/resources/views/components/card/card.blade.php`

Prior to the change if you included `<x-card></x-card>`, it would anonymously render the `index.blade.php` view. After the change, it would use the `Card.php` class.

Although that seems like a very unlikely contrived example. I'll leave it to @taylorotwell to judge if that justifies pushing to 12.x

browner12 commented:
@taylorotwell it definitely looks possible to support the folder name convention for anonymous components. I believe we would just need to adjust the `guessAnonymousComponentUsingNamespaces` and `guessAnonymousComponentUsingPaths` methods. 

It seem unlikely, and very odd that someone would intentionally use both the "index" and "folder" name in their view folder for anonymous components as it doesn't make a ton of sense semantically.

- `/resources/views/components/card/index.blade.php`
- `/resources/views/components/card/card.blade.php`
- `/resources/views/components/card/header.blade.php`
- `/resources/views/components/card/body.blade.php`

However, I *believe* we would still be okay as long as we give the `index` view higher priority than the `card` view in the compiler.  This would cause the `<x-card>` component to still render the `index.blade.php` view if it exists, and only render the `card.blade.php` view when the `index.blade.php` **doesn't** exist.

I will say I *don't* use the alias feature for 3rd party custom components, so I'm not sure if that would be affected any differently.

I would propose we add this feature in a separate PR if you'd like to go forward with it.

taylorotwell commented:
I think I would rather see it in this PR if that's alright - since I wouldn't merge this one without the other change because of the inconsistency.

browner12 committed:
allow using folder name for anonymous index components

per @taylorotwell s request, allow users to name their root components with `index.blade.php` OR the name of the component folder.

browner12 committed:
fix tests

the code changes add an extra `exists()` call, so we'll adjust our numbers here.

browner12 committed:
add some tests

these are copy/paste adjustments of the existing "index" tests.

browner12 commented:
@taylorotwell the anonymous components now support using the folder name instead of `index.blade.php`. so you could structure your accordion like

- `/accordion/accordion.blade.php`
- `/accordion/accordion-item.blade.php`

and then use it like this still:

```blade
<x-accordion>
    <x-accordion.item>Item 1</x-accordion.item>
    <x-accordion.item>Item 2</x-accordion.item>
    <x-accordion.item>Item 3</x-accordion.item>
</x-accordion>
```

the `index.blade.php` view will still take precedence if it is present, and we'll only use the folder name if the `index` is missing.

taylorotwell commented:
Thanks!

royduin commented:
This doesn't seem to work with correctly with vendor prefixes, when I've `<x-prefix::input>` it's working fine with `vendorpackage/resources/components/input/index.blade.php` but it doesn't resolve `vendorpackage/resources/components/input/input.blade.php`. When I add some dumps it should end up with: `prefix::components.input.input` but comes with `prefix::components.input.prefix::input`

More info: https://github.com/rapidez/blade-components/pull/6

browner12 commented:
thanks for catching that @royduin . I was curious how the vendor prefixes would do with this. Luckily it looks like it doesn't break any existing behavior, so we just need to fix it for the new behavior.

royduin commented:
Hey @browner12, do we needs some docs on this? Currently it's undocumented.

browner12 commented:
> Hey @browner12, do we needs some docs on this? Currently it's undocumented.

yup, was just thinking about that today. i'll get on it
